src/components/AccountSeed.js

```diff
@@ -17,154 +17,93 @@
-    const { cursorPosition } = this.state;
-    let wordPosition = 0;
-    let cursor = 0;
-    let char = '';
-    let wasLetter = false;
-    if (0 === value.length) {
-      return 0;
-    }
-    while (true) {
-      if (cursorPosition === cursor || (char = value.charAt(cursor)) === '') {
-        break;
-      }
-      if ([' ', '\n', '\r'].indexOf(char) === -1) {
-        wasLetter = true;
-      } else {
-        if (wasLetter) {
-          ++wordPosition;
-        }
-        wasLetter = false;
-      }
-      ++cursor;
-    }
-    return wordPosition;
-  }
-
-  getSuggestions(input, words) {
-    let fromIndex = WORDS.findIndex(w => w.startsWith(input));
+  getSuggestions = (inputWordArray, word_list) => {
+    const input = inputWordArray[inputWordArray.length - 1]
+    // the word list is sorted, get the index we should start searching
+    // for the word
+    let fromIndex = word_list.findIndex(w => w.startsWith(input));
     if (fromIndex === -1) {
       return [];
     }
-    if (WORDS[fromIndex] === input) {
-      fromIndex = 0;
-    }
+
     const SUGGESTIONS_COUNT = 5;
     const result = [];
     let yielded = 0;
-    result.push(WORDS[fromIndex]);
-    while (yielded < SUGGESTIONS_COUNT && WORDS[fromIndex] !== undefined) {
-      ++fromIndex;
-      if (!WORDS[fromIndex].startsWith(input)) {
+    while (yielded < SUGGESTIONS_COUNT && fromIndex < word_list.length) {
+      if (!word_list[fromIndex].startsWith(input)) {
         return result;
       }
-      if (words.indexOf(WORDS[fromIndex]) !== -1) {
-        continue;
+
+      // do not suggest words that where already used (and part of `inputWordArray`)
+      // unless it's the latest word e.g what the user is typing
+      if (inputWordArray.indexOf(word_list[fromIndex]) === -1 || inputWordArray.indexOf(word_list[fromIndex]) === (inputWordArray.length - 1)) {
+        result.push(word_list[fromIndex]);
+        yielded++
       }
-      result.push(WORDS[fromIndex]);
-      ++yielded;
+      fromIndex++;
     }
     return result;
   }
 
-  renderSuggestions() {
-    const { value } = this.props;
-    const words = value.length ? value.split(' ') : [];
-    const wordPosition = this.getWordPosition();
-    let searchInput = this.getSearchInput();
-    if (WORDS_INDEX[searchInput]) {
-      searchInput = '';
+  narrowWordList = (inputWordArray) => {
+    // if the last word is only present in BIP39 word list
+    if (BIP39_WORDS.indexOf(inputWordArray[inputWordArray.length - 2]) !== -1 && PARITY_WORDS.indexOf(inputWordArray[inputWordArray.length - 2]) === -1) {
+      this.setState({ useParityWordList: false })
+      // if the second last word is only present in Parity word list
+    } else if (BIP39_WORDS.indexOf(inputWordArray[inputWordArray.length - 2]) === -1 && PARITY_WORDS.indexOf(inputWordArray[inputWordArray.length - 2]) !== -1) {
+      this.setState({ useBIP39WordList: false })
     }
-    const suggestions = this.getSuggestions(searchInput, words);
+  }
```

Comment:
Would suggest:

```js
narrowWordList = (inputWordArray) => {
  if (inputWordArray.length < 2) {
    return;
  }

  const previousWord = inputWordArray[inputWordArray.length - 2];

  // Both of these should use binary search!
  const isBIP39 = BIP39_WORDS.indexOf(previousWord) !== -1;
  const isParity = PARITY_WORDS.indexOf(previousWord) !== -1;

  if (isBIP39 && !isParity) {
    this.setState({ useParityWordList: false });
  } else if (!isBIP39 && isParity) {
    this.setState({ useBIP39WordList: false });
  }
}
```

Is there a way to revert this if I enter a word and then delete it after?

Reply:
Thanks, again much easier to read. No way I thought about of to reset the assumption. We could add some state reset if the `inputWords <= 2 &&  (this.state.useBIP39WordList || this.state.useBIP39WordList)`